compiler/typecheck/TcExpr.hs

```diff
@@ -1839,31 +1839,35 @@ tc_infer_id lbl id_name
   where
     return_id id = return (HsVar noExtField (noLoc id), idType id)
 
-    return_data_con con
+    -- TODO restore non-stupid theta part
+    {-
        -- For data constructors, must perform the stupid-theta check
       | null stupid_theta
       = return (HsConLikeOut noExtField (RealDataCon con), con_ty)
+    -}
 
-      | otherwise
+    return_data_con con
        -- See Note [Instantiating stupid theta]
-      = do { let (tvs, theta, rho) = tcSplitSigmaTy con_ty
-           ; (subst, tvs') <- newMetaTyVars tvs
+      = do { let tvs = dataConUserTyVarBinders con
+                 theta = dataConOtherTheta con
+                 args = dataConOrigArgTys con
+                 res = dataConOrigResTy con
+           ; (subst, tvs') <- newMetaTyVars (map binderVar tvs)
            ; let tys'   = mkTyVarTys tvs'
                  theta' = substTheta subst theta
-                 rho'   = substTy subst rho
+                 args'  = map (mapScaledType (substTy subst)) args
+                 res'   = substTy subst res
            ; wrap <- instCall (OccurrenceOf id_name) tys' theta'
-           ; addDataConStupidTheta con (drop (length (fst $ dataConMulVars con)) tys')
-           -- The first K arguments of `tys'` are multiplicities.
-           -- They are followed by the dictionaries which are the stupid
-           -- theta. Thus, we ignore the first K arguments as we just want to
-           -- instantiate dictionary arguments in `addDataConStupidTheta`.
-           -- It might be better to use `dataConRepType` in `con_ty` below.
-           ; return ( mkHsWrap wrap (HsConLikeOut noExtField (RealDataCon con))
-                    , rho') }
-
-      where
-        con_ty         = dataConUserType con
-        stupid_theta   = dataConStupidTheta con
+           -- TODO add a Note on wrappers
+           ; (_subst, mul_vars) <- newMetaTyVars (multiplicityTyVarList (length args') (map getOccName (binderVars tvs)))
```

Comment:
I think that the ultimate game plan here would be to proceed like Richard's eager VTA implementation, and slurp all the arguments provided syntactically, and only eta-expand those arguments which are not already applied. Probably when eager VTA lands.